pyensign/connection.py

```diff
@@ -93,31 +100,35 @@ def __init__(
             self.client_id = client_id
 
     @catch_rpc_error
-    async def publish(self, topic_id, events, on_ack=None, on_nack=None):
+    async def publish(self, topic, events, on_ack=None, on_nack=None):
         if self.shutdown.is_set():
             raise EnsignClientClosingError("client is closing")
 
-        # Create a hash of the topic ID
-        topic_hash = str(topic_id)
+        # Attempt to hash the topic, which fails if there is no topic ID
+        try:
+            topic_hash = hash(topic)
+        except ValueError:
+            topic_hash = None
 
-        # Check if there is already an open stream for this topic
-        if topic_hash in self.publishers:
+        # Check if there is already an open publish stream for this topic
+        if topic_hash and topic_hash in self.publishers:
             publisher = self.publishers[topic_hash]
         else:
             # Create the publish stream for this topic
             publisher = Publisher(
                 self,
-                topic_id,
+                topic,
                 on_ack=on_ack,
                 on_nack=on_nack,
             )
-            self.publishers[topic_hash] = publisher
 
             # Connect to the publish stream
             # TODO: Distinguish between authentication errors, topic errors, and connection errors
             await publisher.connect()
 
             # Run the publisher as a concurrent task which handles reconnects
+            topic_hash = hash(topic)
```

Comment:
I think it might be helpful to explain why we have to reinstantiate `topic_hash` here.